ts/Stock/Indicators/DMI/DMIIndicator.ts

```diff
@@ -0,0 +1,398 @@
+     * @sample stock/indicators/dmi
+     *         DMI indicator
+     *
+     * @extends      plotOptions.sma
+     * @since        next
+     * @product      highstock
+     * @excluding    allAreas, colorAxis, joinBy, keys, navigatorOptions,
+     *               pointInterval, pointIntervalUnit, pointPlacement,
+     *               pointRange, pointStart, showInNavigator, stacking
+     * @requires     stock/indicators/indicators
+     * @requires     stock/indicators/dmi
+     * @optionparent plotOptions.dmi
+     */
+    public static defaultOptions: DMIOptions = merge(SMAIndicator.defaultOptions, {
+        /**
+         * @excluding index
+         */
+        params: {
+            index: void 0, // DMI has unchangeable index
+            period: 14
+        },
+        marker: {
+            enabled: false
+        },
+        tooltip: {
+            pointFormat: '<span style="color:{point.color}">\u25CF</span><b> {series.name}</b><br/>' +
+                '<span style="color:{point.color}">DX</span>: {point.y}<br/>' +
+                '<span style="color:{point.series.options.plusDILine.styles.lineColor}">+DI</span>' +
+                    ': {point.plusDI}<br/>' +
+                '<span style="color:{point.series.options.minusDILine.styles.lineColor}">-DI</span>' +
+                    ': {point.minusDI}<br/>'
+        },
+        /**
+         * +DI line options.
+         */
+        plusDILine: {
+            /**
+             * Styles for the +DI line.
+             */
+            styles: {
+                /**
+                 * Pixel width of the line.
+                 */
+                lineWidth: 1,
+                /**
+                 * Color of the line.
+                 *
+                 * @type {Highcharts.ColorString}
+                 */
+                lineColor: '#00ff00'
+            }
+        },
+        /**
+         * -DI line options.
+         */
+        minusDILine: {
+            /**
+             * Styles for the -DI line.
+             */
+            styles: {
+                /**
+                 * Pixel width of the line.
+                 */
+                lineWidth: 1,
+                /**
+                 * Color of the line.
+                 *
+                 * @type {Highcharts.ColorString}
+                 */
+                lineColor: '#ff0000'
+            }
+        },
+        dataGrouping: {
+            approximation: 'averages'
+        }
+    } as DMIOptions);
+
+    /* *
+     *
+     *  Functions
+     *
+     * */
+
+    public calculateDM(
+        yVal: Array<Array<number>>,
+        i: number,
+        isPositiveDM?: boolean
+    ): number {
+        const currentHigh = yVal[i][1],
+            currentLow = yVal[i][2],
+            previousHigh = yVal[i - 1][1],
+            previousLow = yVal[i - 1][2];
+
+        // If any value is null, return 0.
+        if (!currentHigh || !currentLow || !previousHigh || !previousLow) {
+            return 0;
+        }
```

Comment:
I have a question here. I wanted to handle `null` values for `open` or `low`, but it clearly changes the further indicator values. Any suggestions?

Reply:
I'm afraid none of the indicators is now resolving `null` values in any way. If we decide to do that, then it will be probably resolved in the indicators core, not per indicator (by filtering out null/invalid points/values).